### ISSUE TITLE
Invalid TypeScript definitions generated

### ISSUE THREAD

**Issue by ooflorent** [dotansimha/graphql-code-generator#2699]
**Describe the bug**

Invalid TypeScript definitions are generated.

```ts
export type GetMovieQuery = { node:  };
                                   ^
```

**To Reproduce**

1. My GraphQL schema:

```graphql
type Query {
  node(gid: ID!): Node
}

interface Node {
  gid: ID!
}

interface Entity {
  gid: ID!
  active: Boolean!
} 

type Movie implements Node, Entity {
  gid: ID!
  active: Boolean!
  name: String!
}
```

2. My GraphQL operations:

```graphql
query getMovie($gid: ID!) {
  node(gid: $gid) {
    ... on Entity {
      gid
    }
  }
}
```

3. My `codegen.yml` config file:

```yml
schema: var/schema.json
overwrite: true
generates:
  assets/graphql/types.tsx:
    documents: 'assets/**/*.gql'
    plugins:
      - typescript
      - typescript-operations
    config:
      nonOptionalTypename: false
      skipTypename: true
      preResolveTypes: true
```

**Expected behavior**

`types.tsx` should be a valid TS file.

**Environment:**

- OS: macOS 10.14.6
- `@graphql-codegen/add`: 1.8.1
- `@graphql-codegen/cli`: 1.8.1
- `@graphql-codegen/fragment-matcher`: 1.8.1
- `@graphql-codegen/typescript`: 1.8.1
- `@graphql-codegen/typescript-operations`: 1.8.1
- NodeJS: 12.10.0

**Comment by ooflorent** [comment]
Please note that changing the query to this:

```gql
query getMovie($gid: ID!) {
  node(gid: $gid) {
    gid
    ... on Entity {
      ...SomeEntityFragment
    }
  }
}
```

Results in this:

```ts
export type GetEntityBrandActivationQuery = {
  node:
    | { gid: string }
    | { gid: string }
    | { gid: string }
    | { gid: string }
    | { gid: string }
    | { gid: string }
    | { gid: string }
    | { gid: string }
    | { gid: string }
};
```

Where `{ gid: string }` is produced for each type implementing `Node`.
Moreover, the fields from `SomeEntityFragment` are not generated.

**Comment by ooflorent** [comment]
When using version `1.7.0` the following type is generated (which also is invalid):

```ts
export type GetMovieQuery = { node: {  } | {  } | {  } | {  } | {  } | {  } | {  } | {  } | {  } | {  } | {  } };
```

**Comment by ooflorent** [comment]
I tracked down the bug and it looks like the issue is the 2 interface levels. If I add a new entrypoint:

```gql
type Query {
  entity(gid: String!): Entity!
}
```

And change the query to:

```gql
fragment EntityBrandActivation on Entity {
  gid
  brandData(brand: $brand) {
    active
  }
}

fragment EntityCacheKey on Entity {
  ... on Company {
    id
  }
  ... on Movie {
    id
  }
  ... on Series {
    id
  }
  ... on Person {
    id
  }
  ... on Theater {
    id
  }
}

query getEntityBrandActivation($gid: ID!, $brand: ID!) {
  entity(gid: $gid) {
    ...EntityBrandActivation
    ...EntityCacheKey
  }
}
```

Then the generated types are:

```ts
export type GetEntityBrandActivationQueryVariables = {
  gid: Scalars['ID'],
  brand: Scalars['ID']
};

export type GetEntityBrandActivationQuery = { entity: ({  } | {  } | {  } | {  } | {  } | {  } | {  } | {  } | {  } | {  } | {  })
//                                                     ^^^^
//                                            this is weird
    & EntityBrandActivationFragment
    & EntityCacheKeyFragment
   };
```

**Comment by ooflorent** [comment]
I tried `1.8.2-alpha-a83224e6.9` and a new bug appeared:

```ts
export type GetEntityBrandActivationQuery = { entity: (
    EntityBrandActivation_User_Fragment
    & EntityCacheKey_User_Fragment
  ) | (
    EntityBrandActivation_Company_Fragment
    & EntityCacheKey_Company_Fragment
  ) | (
    EntityBrandActivation_Image_Fragment
    & EntityCacheKey_Image_Fragment
  ) | (
    EntityBrandActivation_Series_Fragment
    & EntityCacheKey_Series_Fragment
  ) | (
    EntityBrandActivation_Person_Fragment
    & EntityCacheKey_Person_Fragment
  ) | (
    EntityBrandActivation_Episode_Fragment
    & EntityCacheKey_Episode_Fragment
  ) …
```

Instead of generating empty objects it now generates fragments. However, **some fragment types are not generated** which results in an error:

> `Cannot find name 'EntityCacheKey_Screen_Fragment' .ts(2304)`

**Comment by dotansimha** [comment]
@ooflorent Any chance you can share a complete schema and queries that reproduces those issues?

**Comment by ooflorent** [comment]
Unfortunately I'm not allowed to share this. I'm sorry… Due to the lack of information, I'd understand if you'd want to close the issue.

**Comment by dotansimha** [comment]
@ooflorent I really want to solve it :) 
Is it possible to share it with `type A {} type B {}`? only with the minimal types that reproduces the error, I think it's happens because some kind of combination of `interface`s and `union`s.

**Comment by dotansimha** [comment]
@ooflorent Can you just share how your types are defined? (for example - `type A implements B`, `interface B`, `union C = D | A` or something similar)

**Comment by dotansimha** [comment]
Managed to reproduce it:

```ts
export type Maybe<T> = T | null;
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string,
  String: string,
  Boolean: boolean,
  Int: number,
  Float: number,
  DateTime: any,
};

export type Company = Element & Node & Entity & {
   __typename?: 'Company',
  active: Scalars['Boolean'],
  createdAt: Scalars['DateTime'],
  createdBy?: Maybe<User>,
  updatedAt: Scalars['DateTime'],
  updatedBy?: Maybe<User>,
  id: Scalars['ID'],
  brandData?: Maybe<EntityBrandData>,
};


export type CompanyBrandDataArgs = {
  brand: Scalars['ID']
};


export type Element = {
  active: Scalars['Boolean'],
  createdAt: Scalars['DateTime'],
  createdBy?: Maybe<User>,
  updatedAt: Scalars['DateTime'],
  updatedBy?: Maybe<User>,
};

export type Entity = {
  brandData?: Maybe<EntityBrandData>,
};


export type EntityBrandDataArgs = {
  brand: Scalars['ID']
};

export type EntityBrandData = {
   __typename?: 'EntityBrandData',
  active: Scalars['Boolean'],
  browsable: Scalars['Boolean'],
  description: Scalars['String'],
  alternateTitle?: Maybe<Scalars['String']>,
  title: Scalars['String'],
};

export type Movie = Element & Node & Entity & {
   __typename?: 'Movie',
  active: Scalars['Boolean'],
  createdAt: Scalars['DateTime'],
  createdBy?: Maybe<User>,
  updatedAt: Scalars['DateTime'],
  updatedBy?: Maybe<User>,
  id: Scalars['ID'],
  brandData?: Maybe<EntityBrandData>,
};


export type MovieBrandDataArgs = {
  brand: Scalars['ID']
};

export type Node = {
  id: Scalars['ID'],
};

export type Query = {
   __typename?: 'Query',
  node: Node,
};


export type QueryNodeArgs = {
  id: Scalars['ID']
};

export type Theater = Element & Node & Entity & {
   __typename?: 'Theater',
  active: Scalars['Boolean'],
  createdAt: Scalars['DateTime'],
  createdBy?: Maybe<User>,
  updatedAt: Scalars['DateTime'],
  updatedBy?: Maybe<User>,
  id: Scalars['ID'],
  brandData?: Maybe<EntityBrandData>,
};


export type TheaterBrandDataArgs = {
  brand: Scalars['ID']
};

export type User = Element & Node & Entity & {
   __typename?: 'User',
  active: Scalars['Boolean'],
  name: Scalars['String'],
  createdAt: Scalars['DateTime'],
  createdBy?: Maybe<User>,
  updatedAt: Scalars['DateTime'],
  updatedBy?: Maybe<User>,
  id: Scalars['ID'],
  brandData?: Maybe<EntityBrandData>,
};


export type UserBrandDataArgs = {
  brand: Scalars['ID']
};
export type GetEntityBrandDataQueryVariables = {
  gid: Scalars['ID'],
  brand: Scalars['ID']
};


export type GetEntityBrandDataQuery = (
  { __typename?: 'Query' }
  & { node: (
    { __typename: 'User' }
    & Pick<User, 'id'>
  ) | (
    { __typename: 'Company' }
    & Pick<Company, 'active' | 'id'>
  ) | (
    { __typename: 'Theater' }
    & Pick<Theater, 'active' | 'id'>
  ) | (
    { __typename: 'Movie' }
    & Pick<Movie, 'id'>
  ) }
);

type EntityBrandData_User_Fragment = (
  { __typename?: 'User' }
  & { brandData: Maybe<(
    { __typename?: 'EntityBrandData' }
    & Pick<EntityBrandData, 'active' | 'browsable' | 'title' | 'alternateTitle' | 'description'>
  )> }
);

type EntityBrandData_Company_Fragment = (
  { __typename?: 'Company' }
  & { brandData: Maybe<(
    { __typename?: 'EntityBrandData' }
    & Pick<EntityBrandData, 'active' | 'browsable' | 'title' | 'alternateTitle' | 'description'>
  )> }
);

type EntityBrandData_Theater_Fragment = (
  { __typename?: 'Theater' }
  & { brandData: Maybe<(
    { __typename?: 'EntityBrandData' }
    & Pick<EntityBrandData, 'active' | 'browsable' | 'title' | 'alternateTitle' | 'description'>
  )> }
);

type EntityBrandData_Movie_Fragment = (
  { __typename?: 'Movie' }
  & { brandData: Maybe<(
    { __typename?: 'EntityBrandData' }
    & Pick<EntityBrandData, 'active' | 'browsable' | 'title' | 'alternateTitle' | 'description'>
  )> }
);

export type EntityBrandDataFragment = EntityBrandData_User_Fragment | EntityBrandData_Company_Fragment | EntityBrandData_Theater_Fragment | EntityBrandData_Movie_Fragment;

type ElementMetadata_User_Fragment = (
  { __typename?: 'User' }
  & Pick<User, 'createdAt' | 'updatedAt'>
  & { createdBy: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name'>
  )>, updatedBy: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name'>
  )> }
);

type ElementMetadata_Company_Fragment = (
  { __typename?: 'Company' }
  & Pick<Company, 'createdAt' | 'updatedAt'>
  & { createdBy: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name'>
  )>, updatedBy: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name'>
  )> }
);

type ElementMetadata_Theater_Fragment = (
  { __typename?: 'Theater' }
  & Pick<Theater, 'createdAt' | 'updatedAt'>
  & { createdBy: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name'>
  )>, updatedBy: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name'>
  )> }
);

type ElementMetadata_Movie_Fragment = (
  { __typename?: 'Movie' }
  & Pick<Movie, 'createdAt' | 'updatedAt'>
  & { createdBy: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name'>
  )>, updatedBy: Maybe<(
    { __typename?: 'User' }
    & Pick<User, 'id' | 'name'>
  )> }
);

export type ElementMetadataFragment = ElementMetadata_User_Fragment | ElementMetadata_Company_Fragment | ElementMetadata_Theater_Fragment | ElementMetadata_Movie_Fragment;

        function test(q: GetEntityBrandDataQuery): void {
          const typeName: 'Company' | 'Theater' | 'User' | 'Movie' = q.node.__typename; // just to check that those are the types we want here
          const brandData = q.node.brandData; // this was missing in the original issue
          const createdAt = q.node.createdAt; // this was missing in the original issue

          if (q.node.__typename === 'Company') {
            console.log('Company:', q.node.active);
          } else if (q.node.__typename === 'Theater') {
            console.log('Company:', q.node.active);
          } else if (q.node.__typename === 'User') {
            console.log('Company:', q.node.id);
          } else if (q.node.__typename === 'Movie') {
            console.log('Company:', q.node.id);
          }
        }
      
```

**Comment by dotansimha** [comment]
@ooflorent I think I managed to fix this issue in: https://github.com/dotansimha/graphql-code-generator/pull/2737

Can you please try the latest alpha? `1.8.2-alpha-9ae02b14.29`

**Comment by dotansimha** [comment]
Fixed in 1.8.2.
